torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,222 @@
+        return x
+    return new_fn
+
+class QuantWrapper(nn.Module):
+    def __init__(self, module, qconfig):
+        super(QuantWrapper, self).__init__()
+        self.quant = QuantStub(qconfig)
+        self.dequant = DeQuantStub()
+        self.module = module
+
+    def forward(self, X):
+        X = self.quant(X)
+        self.module.forward(X)
+        X = self.dequant(X)
+        return X
+
+def get_config_key(name, qconfig_dict):
+    if name in qconfig_dict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
+        return get_config_key(parent, qconfig_dict)
+    else:
+        return '' if '' in qconfig_dict else None
+
+def get_module(mod, name):
+    if name == '':
+        return mod
+    splits = name.split('.')
+    for split in splits:
+        mod = mod._modules[split]
+    return mod
+
+def add_quant_dequant_module(module, qconfig_dict, name):
+    assert name != ''
+    qconfig = qconfig_dict[name]
+    splits = name.split('.')
+    print('splits:', splits)
+    for split in splits:
+        parent_module = module
+        module = module._modules[split]
+    assert parent_module is not module
+    parent_module._modules[splits[-1]] = QuantWrapper(module, qconfig)
+
+def add_quant_dequant(module, qconfig_dict):
+    r"""Add QuantStub and DeQuantStub module and Wrap the module in QuantWrapper
+    according to qconfig_dict
+    """
+    mod_key_set = set()
+    for name, _ in module.named_modules():
+        print('name', name)
+        dict_key = get_config_key(name, qconfig_dict)
+        if dict_key is not None:
+            mod_key_set.add(dict_key)
+
+    # We need to update the modules from inner most
+    # module to outer modules(in decreasing order of number of '.')
+    print('mod_key_set', mod_key_set)
+    mod_key_heap = []
+    for name in mod_key_set:
+        heapq.heappush(mod_key_heap, (-name.count('.'), name))
+    while mod_key_heap:
+        _, name = heapq.heappop(mod_key_heap)
+        if name == '':
+            module = QuantWrapper(module, qconfig_dict[name])
+        else:
+            add_quant_dequant_module(module, qconfig_dict, name)
+    return module
+
+def prepare(module, qconfig_dict):
+    r"""Transform a module to a quantized module according to qconfig_dict,
+    it adds observer and quant dequant stub modules to the module. and
+    changes the forward method to invoke them
+
+    Args:
+        mod: instance of the module we want to transform
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                     configuration
+    """
+    add_observer(module, qconfig_dict)
+    return add_quant_dequant(module, qconfig_dict)
+
+class QuantStub(nn.Module):
+    r"""Quantize stub module which will be replace to actual Quantize module
+    in swap_module function
+    """
+    def __init__(self, qconfig):
+        super(QuantStub, self).__init__()
+        self.add_module('observer', qconfig.activation())
```

Comment:
to continue this discussion - what happens if I manually insert QuantStub in my model? The add_observer would fail since it already has an observer. We probably want to handle QuantStub separately there or we can reverse the order of add_quant_dequant and adding observers

Reply:
Or we could have the add_observer function check if a module already has an observer and do nothing in that case

Reply:
I think it you put quant_dequant addition step (either automatic or manual) before add_observers, then you don't need to have special handling of observer here - it'd be just magically handled by injection code

Reply:
Yes, that is possible: Right now the quantstub has an observer within it, we will need to remove that for it to work.

Reply:
quant_dequant changes the hierarchy so we can't swap the order